src/device/mod.rs

```diff
@@ -553,99 +580,84 @@ impl Device {
-                            // A data packet, with at least a header
-                            let peer_idx = u32::from_le_bytes(make_array(&packet[4..])) >> 8;
-                            peers_by_idx.get(&peer_idx)
-                        }
-                        _ => continue,
+                        Packet::HandshakeResponse(p) => d.peers_by_idx.get(&(p.receiver_idx >> 8)),
+                        Packet::PacketCookieReply(p) => d.peers_by_idx.get(&(p.receiver_idx >> 8)),
+                        Packet::PacketData(p) => d.peers_by_idx.get(&(p.receiver_idx >> 8)),
                     };
 
-                    let peer = match peer {
-                        Some(peer) => peer,
-                        _ => continue,
-                    };
+                    if peer.is_none() {
+                        continue;
+                    }
+
+                    let peer = peer.unwrap();
 
                     // We found a peer, use it to decapsulate the message+
                     let mut flush = false; // Are there packets to send from the queue?
-                    match peer.decapsulate(packet, &mut t.dst_buf[..]) {
+                    match peer
+                        .tunnel
+                        .handle_verified_packet(parsed_packet, &mut t.dst_buf[..])
+                    {
                         TunnResult::Done => {}
-                        TunnResult::Err(e) => {
-                            eprintln!("Decapsulate error {:?}", e);
-                            continue;
-                        }
+                        TunnResult::Err(_) => continue,
                         TunnResult::WriteToNetwork(packet) => {
                             flush = true;
                             peer.add_tx_bytes(udp.sendto(packet, addr));
                         }
                         TunnResult::WriteToTunnelV4(packet, addr) => {
-                            if peer.is_allowed_ip(IpAddr::from(addr)) {
-                                peer.add_rx_bytes(iface.write4(packet))
+                            if peer.is_allowed_ip(addr) {
+                                peer.add_rx_bytes(t.iface.write4(packet))
                             }
                         }
                         TunnResult::WriteToTunnelV6(packet, addr) => {
-                            if peer.is_allowed_ip(IpAddr::from(addr)) {
-                                peer.add_rx_bytes(iface.write6(packet))
+                            if peer.is_allowed_ip(addr) {
+                                peer.add_rx_bytes(t.iface.write6(packet))
                             }
                         }
                     };
 
                     if flush {
                         // Flush pending queue
                         while let TunnResult::WriteToNetwork(packet) =
-                            peer.decapsulate(&[], &mut t.dst_buf[..])
+                            peer.tunnel.decapsulate(None, &[], &mut t.dst_buf[..])
```

Comment:
The only time we call `decapsulate` with `None` for address is when we're flushing packets, shouldn't we directly expose `send_queued_packet` and change the `decapsulate` argument to `src_addr: IpAddr`?

Reply:
ok

Reply:
Actually it is called from ffi where ip is not available